### PR TITLE
Convert a subset of GPU dialect ops to the OpenCL GPU runtime calls

### PR DESCRIPTION
**Pull request by AndreyPavlenko**
Added a new path, that converts the following GPU dialect ops to the corresponding callsof the OpenCL GPU runtime functions (to be implemented later):
  - gpu.alloc, gpu.dealloc, gpu.memcpy and gpu.launch

The first argument of each runtime's function is a pointer to the context structure. This is not a cl_context, this is an execution context, i.e. a single execution of the module's main function. It contains the queue, wait list (in case of out-of-order mode) and someother data, required for the module ops execution. It's expected, that the pointer to the context is passed to the module's main function as the last argument of type memref with zero dims.

For each gpu.launch operation, 2 additional functions are created:
  - getXXXKernel(): returns the kernel pointer, stored in a global variable. If it's NULL, calls createXXXKernel().
  - createXXXKernel(): Calls the runtime's function, that creates a kernel. SPIRV, kernel name, and sizes are passed to the function. The returned pointer is saved in the global var using `llvm.cmpxchg`, to make sure it doesn't overwrite a kernel, created by another thread.

Finally, a destructor function is created, that calls the corresponding runtime's kernel destroy function and passes the pointers, stored in the global vars. This function must be called by themodule owner, when destroying the module.

The kernel is not a cl_kernel, but a runtime's internal structure, that contains a compiledcl_program, preconfigured cl_kernel and other data, required for the kernel execution. The runtime's launch function clones the preconfigured kernel, sets the arguments and enqueues a command to execute the kernel.

Example of code produced by this path:

<details>

```llvm
module @fragment_name {
  llvm.mlir.global internal constant @gcGpuOclKernel_entry_kernel_SPIRV("SPIRV string") {addr_space = 0 : i32}
  llvm.mlir.global internal constant @gcGpuOclKernel_entry_kernel_Name("entry_kernel\00") {addr_space = 0 : i32}
  llvm.mlir.global internal @gcGpuOclKernel_entry_kernel_Ptr() {addr_space = 0 : i32} : !llvm.ptr

  llvm.func @entry(%arg0: !llvm.ptr, %arg1: !llvm.ptr, %arg2: i64, %arg3: i64, %arg4: i64, %arg5: i64, %arg6: i64, %arg7: !llvm.ptr, %arg8: !llvm.ptr, %arg9: i64, %arg10: i64, %arg11: i64, %arg12: i64, %arg13: i64, %arg14: !llvm.ptr, %arg15: !llvm.ptr, %arg16: i64, %arg17: i64, %arg18: i64, %arg19: i64, %arg20: i64, %arg21: !llvm.ptr, %arg22: !llvm.ptr, %arg23: i64) attributes {llvm.emit_c_interface} {
    %0 = llvm.mlir.constant(0 : index) : i64
    %1 = llvm.call @getGcGpuOclKernel_entry_kernel(%arg21) : (!llvm.ptr) -> !llvm.ptr
    llvm.call @gcGpuOclKernelLaunch(%arg21, %1, %arg8, %0, %arg1, %arg15) vararg(!llvm.func<void (ptr, ptr, ...)>) : (!llvm.ptr, !llvm.ptr, !llvm.ptr, i64, !llvm.ptr, !llvm.ptr) -> ()
    llvm.return
  }

  llvm.func internal @getGcGpuOclKernel_entry_kernel(%arg0: !llvm.ptr) -> !llvm.ptr {
    %0 = llvm.mlir.zero : !llvm.ptr
    %1 = llvm.mlir.addressof @gcGpuOclKernel_entry_kernel_Ptr : !llvm.ptr
    %2 = llvm.load %1 : !llvm.ptr -> !llvm.ptr
    %3 = llvm.icmp "eq" %2, %0 : !llvm.ptr
    llvm.cond_br %3, ^bb1, ^bb2
  ^bb1:  // pred: ^bb0
    %4 = llvm.call @createGcGpuOclKernel_entry_kernel(%arg0) : (!llvm.ptr) -> !llvm.ptr
    llvm.return %4 : !llvm.ptr
  ^bb2:  // pred: ^bb0
    llvm.return %2 : !llvm.ptr
  }

  llvm.func internal @createGcGpuOclKernel_entry_kernel(%arg0: !llvm.ptr) -> !llvm.ptr {
    %0 = llvm.mlir.addressof @gcGpuOclKernel_entry_kernel_Ptr : !llvm.ptr
    %1 = llvm.mlir.zero : !llvm.ptr
    %2 = llvm.mlir.constant(1 : i64) : i64
    %3 = llvm.mlir.constant(64 : i64) : i64
    %4 = llvm.mlir.constant(3 : i64) : i64
    %5 = llvm.mlir.constant(4 : i64) : i64
    %6 = llvm.mlir.addressof @gcGpuOclKernel_entry_kernel_SPIRV : !llvm.ptr
    %7 = llvm.mlir.addressof @gcGpuOclKernel_entry_kernel_Name : !llvm.ptr
    %8 = llvm.getelementptr %7[0, 0] : (!llvm.ptr) -> !llvm.ptr, !llvm.array<13 x i8>
    %9 = llvm.getelementptr %6[0, 0] : (!llvm.ptr) -> !llvm.ptr, !llvm.array<3820 x i8>
    %10 = llvm.alloca %4 x i64 : (i64) -> !llvm.ptr
    llvm.store %3, %10 : i64, !llvm.ptr
    %11 = llvm.getelementptr %10[1] : (!llvm.ptr) -> !llvm.ptr, i64
    llvm.store %3, %11 : i64, !llvm.ptr
    %12 = llvm.getelementptr %10[2] : (!llvm.ptr) -> !llvm.ptr, i64
    llvm.store %2, %12 : i64, !llvm.ptr
    %13 = llvm.alloca %4 x i64 : (i64) -> !llvm.ptr
    llvm.store %2, %13 : i64, !llvm.ptr
    %14 = llvm.getelementptr %13[1] : (!llvm.ptr) -> !llvm.ptr, i64
    llvm.store %2, %14 : i64, !llvm.ptr
    %15 = llvm.getelementptr %13[2] : (!llvm.ptr) -> !llvm.ptr, i64
    llvm.store %2, %15 : i64, !llvm.ptr
    %16 = llvm.alloca %5 x i64 : (i64) -> !llvm.ptr
    llvm.store %3, %16 : i64, !llvm.ptr
    %17 = llvm.getelementptr %16[1] : (!llvm.ptr) -> !llvm.ptr, i64
    llvm.store %3, %17 : i64, !llvm.ptr
    %18 = llvm.getelementptr %16[2] : (!llvm.ptr) -> !llvm.ptr, i64
    llvm.store %3, %18 : i64, !llvm.ptr
    %19 = llvm.getelementptr %16[3] : (!llvm.ptr) -> !llvm.ptr, i64
    llvm.store %3, %19 : i64, !llvm.ptr
    %20 = llvm.call @gcGpuOclKernelCreate(%arg0, %9, %8, %10, %13, %5, %16) : (!llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr, i64, !llvm.ptr) -> !llvm.ptr
    %21 = llvm.cmpxchg %0, %1, %20 acq_rel monotonic : !llvm.ptr, !llvm.ptr
    %22 = llvm.extractvalue %21[1] : !llvm.struct<(ptr, i1)> 
    llvm.cond_br %22, ^bb1, ^bb2
  ^bb1:  // pred: ^bb0
    llvm.return %20 : !llvm.ptr
  ^bb2:  // pred: ^bb0
    %23 = llvm.alloca %2 x !llvm.ptr : (i64) -> !llvm.ptr
    llvm.store %20, %23 : !llvm.ptr, !llvm.ptr
    llvm.call @gcGpuOclKernelDestroy(%2, %23) : (i64, !llvm.ptr) -> ()
    %24 = llvm.extractvalue %21[0] : !llvm.struct<(ptr, i1)> 
    llvm.return %24 : !llvm.ptr
  }

  llvm.func @gcGpuOclModuleDestructor() {
    %0 = llvm.mlir.constant(1 : i64) : i64
    %1 = llvm.mlir.addressof @gcGpuOclKernel_entry_kernel_Ptr : !llvm.ptr
    llvm.fence acquire
    %2 = llvm.load %1 : !llvm.ptr -> !llvm.ptr
    %3 = llvm.alloca %0 x !llvm.ptr : (i64) -> !llvm.ptr
    llvm.store %2, %3 : !llvm.ptr, !llvm.ptr
    llvm.call @gcGpuOclKernelDestroy(%0, %3) : (i64, !llvm.ptr) -> ()
    llvm.return
  }

  llvm.func @gcGpuOclKernelCreate(!llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr, i64, !llvm.ptr) -> !llvm.ptr
  llvm.func @gcGpuOclKernelDestroy(i64, !llvm.ptr)
  llvm.func @gcGpuOclKernelLaunch(!llvm.ptr, !llvm.ptr, ...)
}
```

</details>

### CODE REVIEW

File: include/gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h
Comment (kurapov-peter) [inline include/gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h:18]: Can these be constrexpr strings?

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (kurapov-peter) [inline lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:50]: Where does -3 come from?

**Review by kurapov-peter — COMMENTED** [review]
Could you please add lits? Would help reviewing.

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:50]: The context is passed as the last arg of the function of type memref<any> with zero dims, it is converted to ptr, ptr, i64.

Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:50]: Added a comment.

File: include/gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h
Comment (AndreyPavlenko) [reply include/gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h]: Done.

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (dchigarev) [inline lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:76]: don't we need to dealloc it after the `funcCall`?

File: include/gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h
Comment (AndreyPavlenko) [reply include/gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h:12]: I can add the signatures in comments. It's not expected to have exported C functions. The functions are not public and are injected into the engine when building the module. This allows to get rid of building a shared lib with the exported C functions.

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:44]: A new argument is added to the initial function, before running the passes. We can add a single pointer, but:
1. in this case we need to also change the function call in _mlir_ciface_entry(), that is created by the pipeline.
2. it does not make sense, because all args to the function are passed through the function wrapper and are wrapped into the https://mlir.llvm.org/docs/TargetLLVMIR/#c-compatible-wrapper-emission structure, that has minimum 3 fields.

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:76]: It's allocated on the stack - https://llvm.org/docs/LangRef.html#alloca-instruction

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (kurapov-peter) [inline lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:99]: How is this different from `ConvertAllocOpToGpuRuntimeCallPattern` in handling the memrefs and shapes? I'd expect the lowering to look just the same except for the runtime call signature/params.

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (kurapov-peter) [inline lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:502]: Not sure I follow, why do we destroy the kernel here?

File: lib/gc/Transforms/GPU/AddContextArg.cpp
Comment (kurapov-peter) [inline lib/gc/Transforms/GPU/AddContextArg.cpp:34]: Does this mean this wouldn't work in case of more than one block?

File: include/gc/Transforms/Passes.td
Comment (kurapov-peter) [inline include/gc/Transforms/Passes.td:97]: Should it be applied to kernels only?

Comment (AndreyPavlenko) [reply include/gc/Transforms/Passes.td:97]: No, it should be applied to the functions, that launch the kernels. The argument is passed to the runtime.

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:36]: It has the external linkage by default.

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:502]: If the current value of the global field is not null, then some other thread has already created and assigned the the value. In this case, we destroy the created kernel and return that one, created by another thread.

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:99]: ConvertAllocOpToGpuRuntimeCallPattern didn't work for me. Check this test:

```LLVM
module attributes {gpu.container_module} {
  func.func @main() attributes {llvm.emit_c_interface} {
    %0 = "gpux.create_stream"() : () -> !gpux.StreamType
    %memref = "gpux.alloc"(%0) {operandSegmentSizes = array<i32: 0, 1, 0, 0>} : (!gpux.StreamType) -> memref<8xf32>
    return
  }
}
```

`gc-opt --convert-gpux-to-llvm` gives the following result:
```LLVM
module attributes {gpu.container_module} {
  func.func @main() attributes {llvm.emit_c_interface} {
    %0 = llvm.mlir.zero : !llvm.ptr
    %1 = llvm.mlir.zero : !llvm.ptr
    %2 = llvm.call @gpuCreateStream(%0, %1) : (!llvm.ptr, !llvm.ptr) -> !llvm.ptr
    %3 = llvm.mlir.constant(8 : index) : i64
    %4 = llvm.mlir.constant(1 : index) : i64
    %5 = llvm.mlir.zero : !llvm.ptr
    %6 = llvm.getelementptr %5[%3] : (!llvm.ptr, i64) -> !llvm.ptr, f32
    %7 = llvm.ptrtoint %6 : !llvm.ptr to i64
    %8 = llvm.mlir.constant(64 : i64) : i64
    %9 = llvm.mlir.constant(0 : i32) : i32
    %10 = llvm.call @gpuMemAlloc(%2, %7, %8, %9) : (!llvm.ptr, i64, i64, i32) -> !llvm.ptr
    %11 = llvm.mlir.undef : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)>
    %12 = llvm.bitcast %10 : !llvm.ptr to !llvm.ptr
    %13 = llvm.insertvalue %12, %11[0] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    %14 = llvm.bitcast %10 : !llvm.ptr to !llvm.ptr
    %15 = llvm.insertvalue %14, %13[1] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    %16 = llvm.mlir.constant(0 : i64) : i64
    %17 = llvm.insertvalue %16, %15[2] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    %18 = llvm.insertvalue %3, %17[3, 0] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    %19 = llvm.insertvalue %4, %18[4, 0] : !llvm.struct<(ptr, ptr, i64, array<1 x i64>, array<1 x i64>)> 
    return
  }
  llvm.func @gpuCreateStream(!llvm.ptr, !llvm.ptr) -> !llvm.ptr
  llvm.func @gpuMemAlloc(!llvm.ptr, i64, i64, i32) -> !llvm.ptr
}
```
Pay attention to these lines:
```LLVM
    %5 = llvm.mlir.zero : !llvm.ptr
    %6 = llvm.getelementptr %5[%3] : (!llvm.ptr, i64) -> !llvm.ptr, f32
```

File: include/gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h
Comment (AndreyPavlenko) [reply include/gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h:27]: This is a placeholder for #343. It also undefs this one https://github.com/intel/graph-compiler/blob/1610c78b68b8e4bdf73ef1780ded8febf449cc4f/lib/gc/Transforms/GPU/GpuToGpuOcl.cpp#L10

File: lib/gc/Transforms/IterativeTilingAndFusion.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/IterativeTilingAndFusion.cpp]: Worked around by increasing the shapes.

File: test/mlir/test/gc/gpu-runner/gpu-to-gpuocl.mlir
Comment (AndreyPavlenko) [reply test/mlir/test/gc/gpu-runner/gpu-to-gpuocl.mlir:1]: Done.

File: lib/gc/Transforms/GPU/GpuToGpuOcl.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/GpuToGpuOcl.cpp:123]: Now it's almost the same, except the static size calculation.

File: lib/gc/Transforms/GPU/AddContextArg.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/AddContextArg.cpp]: Removed this code.

**Review by kurapov-peter — APPROVED** [review]

File: lib/gc/Transforms/GPU/Pipeline.cpp
Comment (dchigarev) [inline lib/gc/Transforms/GPU/Pipeline.cpp:113]: maybe we should postpone changes to the pipeline until (#343) is merged? Otherwise we're breaking a working pipeline

**Review by dchigarev — COMMENTED** [review]
LGTM

good to merge once [this](https://github.com/intel/graph-compiler/pull/333/files#r1773435834) is addressed

File: test/mlir/test/gc/Transforms/GPU/lit.local.cfg
Comment (dchigarev) [inline test/mlir/test/gc/Transforms/GPU/lit.local.cfg:5]: The test should pass even without #343 since it's just the lowering, right?

Comment (AndreyPavlenko) [reply test/mlir/test/gc/Transforms/GPU/lit.local.cfg]: You are right, it should pass.

File: lib/gc/Transforms/GPU/Pipeline.cpp
Comment (AndreyPavlenko) [reply lib/gc/Transforms/GPU/Pipeline.cpp]: Reverted changes to the pipeline.

**Review by dchigarev — APPROVED** [review]
good to be merged from my side

**Pull request merged**
